File: src/controllers/taskController.js
const taskService = require('../services/taskService');
const auditService = require('../services/auditService');
const Notification = require('../models/Notification');
const { emitToUser, emitToProject, emitToTask } = require('../services/socketService');
const { parseMentions, resolveMentions, notifyMentions } = require('../services/mentionService');

/**
 * POST /api/v1/tasks
 */
const createTask = async (req, res, next) => {
  try {
    const task = await taskService.createTask(req.body, req.user._id);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'TASK_CREATED',
      entity: 'Task',
      entityId: task._id,
      description: `Task created: "${task.title}"`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    // Real-time: notify assigned user
    if (task.assignedTo && task.assignedTo._id && task.assignedTo._id.toString() !== req.user._id.toString()) {
      // Create persistent notification
      const notif = await Notification.create({
        user: task.assignedTo._id,
        type: 'TASK_ASSIGNED',
        message: `You have been assigned to task "${task.title}" by ${req.user.name}`,
        relatedEntity: { type: 'Task', id: task._id },
      });
      emitToUser(task.assignedTo._id, 'notification', {
        type: 'TASK_ASSIGNED',
        message: notif.message,
        entityType: 'Task',
        entityId: task._id,
        timestamp: notif.createdAt,
      });
    }
    if (task.project && task.project._id) {
      emitToProject(task.project._id, 'project:task:created', { task });
    }

    // Parse and handle @mentions in description
    if (task.description) {
      const mentionUsernames = parseMentions(task.description);
      if (mentionUsernames.length > 0) {
        const mentionedUsers = await resolveMentions(mentionUsernames);
        if (mentionedUsers.length > 0) {
          await notifyMentions(mentionedUsers, {
            entityType: 'Task',
            entityId: task._id,
            entityTitle: task.title,
            actorId: req.user._id,
            actorName: req.user.name,
          });
        }
      }
    }

    res.status(201).json({
      success: true,
      message: 'Task created successfully',
      data: { task },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/tasks
 */
const getTasks = async (req, res, next) => {
  try {
    const filters = {
      status: req.query.status,
      priority: req.query.priority,
      category: req.query.category,
      project: req.query.project,
      assignedTo: req.query.assignedTo,
      createdBy: req.query.createdBy,
      search: req.query.search,
      overdue: req.query.overdue,
      sortBy: req.query.sortBy,
      sortOrder: req.query.sortOrder,
      label: req.query.label,
      hasDependencies: req.query.hasDependencies,
      userRole: req.user.role,
    };

    const result = await taskService.getTasks(filters, req.pagination, req.user._id);

    res.json({
      success: true,
      message: 'Tasks retrieved successfully',
      data: result.data,
      pagination: result.pagination,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/tasks/:id
 */
const getTask = async (req, res, next) => {
  try {
    const task = await taskService.getTaskById(req.params.id);
    res.json({
      success: true,
      message: 'Task retrieved successfully',
      data: { task },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * PUT /api/v1/tasks/:id
 */
const updateTask = async (req, res, next) => {
  try {
    const task = await taskService.updateTask(req.params.id, req.body, req.user._id, req.user.role);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'TASK_UPDATED',
      entity: 'Task',
      entityId: task._id,
      description: `Task updated: "${task.title}"`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    // Real-time: broadcast task update
    emitToTask(req.params.id, 'task:updated', { task });
    if (task.project && task.project._id) {
      emitToProject(task.project._id, 'project:task:updated', { task });
    }
    if (task.assignedTo && task.assignedTo._id) {
      emitToUser(task.assignedTo._id, 'task:updated', { task });
    }

    // Handle @mentions in description
    if (task.description) {
      const mentionUsernames = parseMentions(task.description);
      if (mentionUsernames.length > 0) {
        const mentionedUsers = await resolveMentions(mentionUsernames);
        if (mentionedUsers.length > 0) {
          await notifyMentions(mentionedUsers, {
            entityType: 'Task',
            entityId: task._id,
            entityTitle: task.title,
            actorId: req.user._id,
            actorName: req.user.name,
          });
        }
      }
    }

    res.json({
      success: true,
      message: 'Task updated successfully',
      data: { task },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * DELETE /api/v1/tasks/:id
 */
const deleteTask = async (req, res, next) => {
  try {
    const result = await taskService.deleteTask(req.params.id, req.user._id, req.user.role);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'TASK_DELETED',
      entity: 'Task',
      entityId: req.params.id,
      description: result.message,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    // Real-time: broadcast task deletion
    emitToTask(req.params.id, 'task:deleted', { taskId: req.params.id });

    res.json({
      success: true,
      message: result.message,
      data: null,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * PATCH /api/v1/tasks/:id/status
 */
const updateStatus = async (req, res, next) => {
  try {
    const { status } = req.body;
    const task = await taskService.updateTask(req.params.id, { status }, req.user._id, req.user.role);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'TASK_STATUS_CHANGED',
      entity: 'Task',
      entityId: task._id,
      description: `Task status changed to: ${status}`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    // Real-time: broadcast status change
    emitToTask(req.params.id, 'task:status', { taskId: req.params.id, status, task });
    if (task.project && task.project._id) {
      emitToProject(task.project._id, 'project:task:status', { taskId: req.params.id, status, task });
    }

    // Create notification for task creator when status changes
    const creatorId = task.createdBy?._id || task.createdBy;
    if (creatorId && creatorId.toString() !== req.user._id.toString()) {
      let notifType = 'TASK_COMPLETED';
      let notifMsg = `Task "${task.title}" status changed to ${status.replace(/_/g, ' ').toLowerCase()} by ${req.user.name}`;
      if (status === 'COMPLETED') {
        notifMsg = `Task "${task.title}" has been completed by ${req.user.name}`;
      } else if (status === 'CANCELLED') {
        notifType = 'TASK_OVERDUE';
        notifMsg = `Task "${task.title}" has been cancelled by ${req.user.name}`;
      }

      const notif = await Notification.create({
        user: creatorId,
        type: notifType,
        message: notifMsg,
        relatedEntity: { type: 'Task', id: task._id },
      });

      emitToUser(creatorId, 'notification', {
        type: notifType,
        message: notif.message,
        entityType: 'Task',
        entityId: task._id,
        timestamp: notif.createdAt,
      });
    }

    res.json({
      success: true,
      message: 'Task status updated successfully',
      data: { task },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * PATCH /api/v1/tasks/:id/assign
 */
const assignTask = async (req, res, next) => {
  try {
    const { assignedTo } = req.body;
    const task = await taskService.assignTask(req.params.id, assignedTo, req.user._id, req.user.role);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'TASK_ASSIGNED',
      entity: 'Task',
      entityId: task._id,
      description: `Task assigned to user`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    // Notify the assigned user
    if (task.assignedTo && task.assignedTo._id && task.assignedTo._id.toString() !== req.user._id.toString()) {
      const notif = await Notification.create({
        user: task.assignedTo._id,
        type: 'TASK_ASSIGNED',
        message: `You have been assigned to task "${task.title}" by ${req.user.name}`,
        relatedEntity: { type: 'Task', id: task._id },
      });
      emitToUser(task.assignedTo._id, 'notification', {
        type: 'TASK_ASSIGNED',
        message: notif.message,
        entityType: 'Task',
        entityId: task._id,
        timestamp: notif.createdAt,
      });
    }

    res.json({
      success: true,
      message: 'Task assigned successfully',
      data: { task },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * PATCH /api/v1/tasks/reorder — kanban drag-and-drop
 */
const reorderTasks = async (req, res, next) => {
  try {
    const result = await taskService.reorderTasks(req.body.updates);
    // Real-time: broadcast reorder to all connected users
    emitToTask('global', 'kanban:reordered', { updates: req.body.updates });
    res.json({
      success: true,
      message: result.message,
      data: { count: result.count },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/tasks/kanban — kanban board data
 */
const getKanbanBoard = async (req, res, next) => {
  try {
    const board = await taskService.getKanbanBoard({
      project: req.query.project,
      assignedTo: req.query.assignedTo,
    }, req.user._id, req.user.role);

    res.json({
      success: true,
      message: 'Kanban board retrieved successfully',
      data: board,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/tasks/calendar — calendar view data
 */
const getCalendarData = async (req, res, next) => {
  try {
    const tasks = await taskService.getCalendarData({
      dateFrom: req.query.dateFrom,
      dateTo: req.query.dateTo,
      project: req.query.project,
    }, req.user._id, req.user.role);

    res.json({
      success: true,
      message: 'Calendar data retrieved successfully',
      data: tasks,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/tasks/dashboard/stats
 */
const getDashboardStats = async (req, res, next) => {
  try {
    const stats = await taskService.getDashboardStats(req.user._id, req.user.role);
    res.json({
      success: true,
      message: 'Dashboard statistics retrieved successfully',
      data: { stats },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/tasks/:id/subtasks
 */
const getSubtasks = async (req, res, next) => {
  try {
    const result = await taskService.getSubtasks(req.params.id, req.user._id);
    res.json({
      success: true,
      message: 'Subtasks retrieved successfully',
      data: result.data,
      total: result.total,
      completed: result.completed,
      percentage: result.percentage,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * POST /api/v1/tasks/:id/subtasks
 */
const createSubtask = async (req, res, next) => {
  try {
    const subtask = await taskService.createSubtask(req.params.id, req.body, req.user._id);
    res.status(201).json({
      success: true,
      message: 'Subtask created successfully',
      data: { subtask },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * PUT /api/v1/tasks/subtasks/:subtaskId
 */
const updateSubtask = async (req, res, next) => {
  try {
    const subtask = await taskService.updateSubtask(req.params.subtaskId, req.body, req.user._id);
    res.json({
      success: true,
      message: 'Subtask updated successfully',
      data: { subtask },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * DELETE /api/v1/tasks/subtasks/:subtaskId
 */
const deleteSubtask = async (req, res, next) => {
  try {
    const result = await taskService.deleteSubtask(req.params.subtaskId, req.user._id);
    res.json({
      success: true,
      message: result.message,
      data: null,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  createTask,
  getTasks,
  getTask,
  updateTask,
  deleteTask,
  updateStatus,
  assignTask,
  reorderTasks,
  getKanbanBoard,
  getCalendarData,
  getDashboardStats,
  getSubtasks,
  createSubtask,
  updateSubtask,
  deleteSubtask,
};

File: backend/src/main/java/com/taskflow/controller/TaskController.java
package com.taskflow.controller;

import com.taskflow.dto.request.TaskRequests;
import com.taskflow.dto.response.ServiceResults;
import com.taskflow.entity.Subtask;
import com.taskflow.entity.Task;
import com.taskflow.entity.User;
import com.taskflow.service.TaskService;
import com.taskflow.util.ApiResponse;
import com.taskflow.util.PaginationUtils;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/tasks")
public class TaskController {

    private final TaskService taskService;

    public TaskController(TaskService taskService) {
        this.taskService = taskService;
    }

    // ─── Dashboard / kanban / calendar (declared before /{id}) ───────────────

    @GetMapping("/dashboard/stats")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getDashboardStats(
            @AuthenticationPrincipal User currentUser) {
        Map<String, Object> stats = taskService.getDashboardStats(currentUser);
        return ResponseEntity.ok(ApiResponse.ok("Dashboard statistics retrieved successfully",
                Map.of("stats", stats)));
    }

    @GetMapping("/kanban")
    public ResponseEntity<ApiResponse<Map<String, List<Task>>>> getKanbanBoard(
            @RequestParam(required = false) String project,
            @RequestParam(required = false) String assignedTo,
            @AuthenticationPrincipal User currentUser) {
        Map<String, List<Task>> board = taskService.getKanbanBoard(project, assignedTo, currentUser);
        return ResponseEntity.ok(ApiResponse.ok("Kanban board retrieved successfully", board));
    }

    @GetMapping("/calendar")
    public ResponseEntity<ApiResponse<List<Task>>> getCalendarData(
            @RequestParam(required = false) String dateFrom,
            @RequestParam(required = false) String dateTo,
            @RequestParam(required = false) String project,
            @AuthenticationPrincipal User currentUser) {
        List<Task> tasks = taskService.getCalendarData(dateFrom, dateTo, project, currentUser);
        return ResponseEntity.ok(ApiResponse.ok("Calendar data retrieved successfully", tasks));
    }

    @PatchMapping("/reorder")
    public ResponseEntity<ApiResponse<Map<String, Object>>> reorderTasks(
            @Valid @RequestBody TaskRequests.ReorderRequest request) {
        int count = taskService.reorderTasks(request.updates());
        return ResponseEntity.ok(ApiResponse.ok("Tasks reordered successfully", Map.of("count", count)));
    }

    // ─── CRUD ────────────────────────────────────────────────────────────────

    @GetMapping
    public ResponseEntity<ApiResponse<List<Task>>> getTasks(@RequestParam Map<String, String> params,
                                                            @AuthenticationPrincipal User currentUser) {
        PaginationUtils.PageRequest page = PaginationUtils.resolve(
                parseInt(params.get("page")), parseInt(params.get("limit")));
        ServiceResults.PageResult<Task> result = taskService.getTasks(params, page, currentUser);
        return ResponseEntity.ok(ApiResponse.ok("Tasks retrieved successfully", result.data(),
                result.pagination()));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Map<String, Object>>> createTask(
            @Valid @RequestBody TaskRequests.CreateTaskRequest request,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Task task = taskService.createTask(request, currentUser, http.getRemoteAddr(), http.getHeader("User-Agent"));
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok("Task created successfully", Map.of("task", task)));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getTask(@PathVariable Long id) {
        return ResponseEntity.ok(ApiResponse.ok("Task retrieved successfully",
                Map.of("task", taskService.getTaskById(id))));
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> updateTask(
            @PathVariable Long id, @RequestBody Map<String, Object> body,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Task task = taskService.updateTask(id, body, currentUser, http.getRemoteAddr(), http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Task updated successfully", Map.of("task", task)));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> deleteTask(@PathVariable Long id,
                                                        @AuthenticationPrincipal User currentUser,
                                                        HttpServletRequest http) {
        taskService.deleteTask(id, currentUser, http.getRemoteAddr(), http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Task deleted successfully", null));
    }

    @PatchMapping("/{id}/status")
    public ResponseEntity<ApiResponse<Map<String, Object>>> updateStatus(
            @PathVariable Long id, @Valid @RequestBody TaskRequests.StatusUpdateRequest request,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Task task = taskService.updateStatus(id, request.status(), currentUser, http.getRemoteAddr(),
                http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Task status updated successfully", Map.of("task", task)));
    }

    @PatchMapping("/{id}/assign")
    public ResponseEntity<ApiResponse<Map<String, Object>>> assignTask(
            @PathVariable Long id, @RequestBody(required = false) TaskRequests.AssignRequest request,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Task task = taskService.assignTask(id, request == null ? null : request.assignedTo(), currentUser,
                http.getRemoteAddr(), http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Task assigned successfully", Map.of("task", task)));
    }

    // ─── Subtasks ────────────────────────────────────────────────────────────

    @GetMapping("/{id}/subtasks")
    public ResponseEntity<ApiResponse<List<Subtask>>> getSubtasks(@PathVariable Long id) {
        ServiceResults.SubtaskProgress progress = taskService.getSubtasks(id);
        return ResponseEntity.ok(ApiResponse.ok("Subtasks retrieved successfully", progress.data())
                .subtaskProgress(progress.total(), progress.completed(), progress.percentage()));
    }

    @PostMapping("/{id}/subtasks")
    public ResponseEntity<ApiResponse<Map<String, Object>>> createSubtask(
            @PathVariable Long id, @Valid @RequestBody TaskRequests.CreateSubtaskRequest request,
            @AuthenticationPrincipal User currentUser) {
        Subtask subtask = taskService.createSubtask(id, request.title(), currentUser);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok("Subtask created successfully", Map.of("subtask", subtask)));
    }

    @PutMapping("/subtasks/{subtaskId}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> updateSubtask(
            @PathVariable Long subtaskId, @RequestBody Map<String, Object> body) {
        Subtask subtask = taskService.updateSubtask(subtaskId, body);
        return ResponseEntity.ok(ApiResponse.ok("Subtask updated successfully", Map.of("subtask", subtask)));
    }

    @DeleteMapping("/subtasks/{subtaskId}")
    public ResponseEntity<ApiResponse<Void>> deleteSubtask(@PathVariable Long subtaskId) {
        taskService.deleteSubtask(subtaskId);
        return ResponseEntity.ok(ApiResponse.ok("Subtask deleted successfully", null));
    }

    private Integer parseInt(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        try {
            return Integer.valueOf(value.trim());
        } catch (NumberFormatException ex) {
            return null;
        }
    }
}
